File: source-code/valuations/discounted-future-market-cap.js
/*
    Model: Discounted Future Market Cap
    
    © Copyright: 
        Discounting Cash Flows Inc. (discountingcashflows.com)
        8 The Green, Dover, DE 19901
*/

Input(
  {
    _DISCOUNT_RATE: '',
    _MARKET_PREMIUM: '',
    _RISK_FREE_RATE: '',
    BETA: '',
    PE_RATIO: '',
    PROJECTION_YEARS: 5,
    HISTORICAL_YEARS: 10,
    _REVENUE_GROWTH_RATE: '',
    _NET_INCOME_MARGIN: '',
  },
  [{
    parent:'_DISCOUNT_RATE',
    children:['BETA', '_RISK_FREE_RATE', '_MARKET_PREMIUM']
  }]
); 

$.when(
  get_income_statement(),
  get_income_statement_ltm(),
  get_balance_sheet_statement(),
  get_balance_sheet_statement_quarterly(2),
  get_quote(),
  get_profile(),
  get_treasury(),
  get_fx(),
  get_risk_premium()).done(
  function($income, $income_ltm, $balance, $balance_quarterly, $quote, $profile, $treasury, $fx, $risk_premium){
  try{
    var response = new Response({
        income: $income,
        income_ltm: $income_ltm,
        balance: $balance,
        balance_quarterly: $balance_quarterly,
        balance_ltm: 'balance_quarterly:0',
        quote: $quote,
        profile: $profile,
        treasury: $treasury,
        risk_premium: $risk_premium,
    }).toOneCurrency('income', $fx).merge('_ltm');
    response.balance[0]['date'] = 'LTM';
    
    // +---------------- ASSUMPTIONS SECTION -----------------+ 
	setAssumption('PE_RATIO', response.quote.pe);
    setAssumption('_MARKET_PREMIUM', response.risk_premium.totalEquityRiskPremium );
    // Risk free rate is the yield of the 10 year treasury note
	setAssumption('_RISK_FREE_RATE', response.treasury.year10);
    // Set beta (used in calculating the discount rate)
    if(response.profile.beta){
    	setAssumption('BETA', response.profile.beta);
    }
    else{
    	setAssumption('BETA', 1);
    }
    var costOfEquity = getAssumption('_RISK_FREE_RATE') + getAssumption('BETA')*getAssumption('_MARKET_PREMIUM');
    setAssumption('_DISCOUNT_RATE', toP(costOfEquity));
    
     // Setup Original Data
    var original_data = new DateValueData({
        revenue: new DateValueList(response.income, 'revenue'),
        costOfRevenue: new DateValueList(response.income, 'costOfRevenue'),
        grossProfit: new DateValueList(response.income, 'grossProfit'),
        netIncome: new DateValueList(response.income, 'netIncome'),
        weightedAverageShsOut: new DateValueList(response.income, 'weightedAverageShsOut'),
        operatingIncome: new DateValueList(response.income, 'operatingIncome'),
        incomeBeforeTax: new DateValueList(response.income, 'incomeBeforeTax'),
        incomeTaxExpense: new DateValueList(response.income, 'incomeTaxExpense'),
        
        totalNonCurrentAssets: new DateValueList(response.balance, 'totalNonCurrentAssets'),
        totalCurrentAssets: new DateValueList(response.balance, 'totalCurrentAssets'),
        totalCurrentLiabilities: new DateValueList(response.balance, 'totalCurrentLiabilities'),
        cashAndCashEquivalents: new DateValueList(response.balance, 'cashAndCashEquivalents'),
        totalStockholdersEquity: new DateValueList(response.balance, 'totalStockholdersEquity'),
    });
    
    var currentDate = original_data.lastDate();
    var nextYearDate = currentDate + 1;
    var startDate = nextYearDate - getAssumption('HISTORICAL_YEARS');
    var forecastEndDate = currentDate + getAssumption('PROJECTION_YEARS');
    
    // Compute historical values and ratios
    var historical_computed_data = original_data.setFormula({
        _grossMargin: ['grossProfit:0', '/', 'revenue:0'],
        _operatingMargin: ['operatingIncome:0', '/', 'revenue:0'],
        _netMargin: ['netIncome:0', '/', 'revenue:0'],
        _revenueGrowthRate: ['function:growth_rate', 'revenue'],
        _netIncomeGrowthRate: ['function:growth_rate', 'netIncome'],
        _totalStockholdersEquityGrowthRate: ['function:growth_rate', 'totalStockholdersEquity'],
        
        // Calculating NOPAT (Net Operating Profit After Taxes)
        _taxRate: ['incomeTaxExpense', '/', 'incomeBeforeTax'],
        incomeTax: ['operatingIncome', '*', '_taxRate'],
        nopat: ['operatingIncome', '-', 'incomeTax'],
        
        // Calculating Invested Capital
        investedCapital: ['function:sum', {'keys': [
            ['totalNonCurrentAssets', '-', 'cashAndCashEquivalents'],
            ['totalCurrentAssets', '-', 'totalCurrentLiabilities'],
        ]}],
        _roic: ['nopat', '/', 'investedCapital'],
    }).compute();
    
	setAssumption('_NET_INCOME_MARGIN', toP(historical_computed_data.get('_netMargin').sublist(startDate).average()));
	setAssumption('_REVENUE_GROWTH_RATE', toP(historical_computed_data.get('_revenueGrowthRate').sublist(startDate).average()));
    
    // Compute forecasted values and ratios
    var forecasted_data = historical_computed_data.removeDate('LTM').setFormula({
      linearRegressionRevenue: ['function:linear_regression', 'revenue', {slope: 1, start_date: startDate}],
      revenue: ['function:compound', 'linearRegressionRevenue:start_date', {rate: getAssumption('_REVENUE_GROWTH_RATE'), start_date: nextYearDate}],
      _revenueGrowthRate: ['function:growth_rate', 'revenue'],
      netIncome: ['revenue:0', '*', getAssumption('_NET_INCOME_MARGIN')],
    }).setEditable(_edit(), {
      start_date: nextYearDate,
      keys: ['revenue', 'netIncome'],
    }).compute({'forecast_end_date': forecastEndDate}); 
    // +------------- END OF ASSUMPTIONS SECTION -------------+
    
    // +---------------- MODEL VALUES SECTION ----------------+
    var projectedRevenue = forecasted_data.get('revenue').lastValue();
    var projectedNetIncome = forecasted_data.get('netIncome').lastValue();
	
    var sharesOutstanding = original_data.get('weightedAverageShsOut').valueAtDate('LTM');
    var futureMarketCap = getAssumption('PE_RATIO') * projectedNetIncome;
    var discountedFutureMarketCap = futureMarketCap / Math.pow(1 + getAssumption('_DISCOUNT_RATE'), getAssumption('PROJECTION_YEARS'));
    var presentValue = discountedFutureMarketCap / sharesOutstanding;
	var currency = response.currency;
	
    // If we are calculating the value per share for a watch, we can stop right here.
    if(_StopIfWatch(presentValue, currency)){
      return;
    }
    _SetEstimatedValue(presentValue, currency);
    
    print(presentValue, 'Present Value', '#', currency);
    print(projectedNetIncome, 'Estimated Net Income in ' + forecastEndDate, '#', currency);
    print(futureMarketCap, 'Estimated Market Capitalisation in ' + forecastEndDate, '#', currency);
    print(discountedFutureMarketCap, 'Market Capitalisation discounted to present', '#', currency);
    print(sharesOutstanding, 'Shares Outstanding', '#');
    // +------------- END OF MODEL VALUES SECTION ------------+
	
    // +------------------- CHARTS SECTION -------------------+
    forecasted_data.renderChart({
      start_date: startDate,
      keys: ['revenue', 'netIncome', 'linearRegressionRevenue'],
      properties: {
        title: 'Historical and forecasted data',
        currency: currency,
      	number_format: 'M',
        disabled_keys: ['linearRegressionRevenue']
      }
    });
	// +---------------- END OF CHARTS SECTION ---------------+ 
	
    // +------------------- TABLES SECTION -------------------+
    // Estimated Future Data
    forecasted_data.renderTable({
      start_date: currentDate,
      data: {
          'Revenue': 'revenue',
          '{%} Revenue Growth Rate': '_revenueGrowthRate',
          'Net Income': 'netIncome',
      },
      properties: {
        'title': 'Estimated Future Values',
        'currency': currency,
        'column_order': 'ascending',
      	'number_format': 'M',
      }
    });
    
    // Historical values table
    historical_computed_data.renderTable({
      start_date: startDate,
      data: {
          'Revenue': 'revenue',
          '{%} Revenue Growth Rate': '_revenueGrowthRate',
          'Cost of Revenue': 'costOfRevenue',
          'Gross Profit': 'grossProfit',
          '{%} Gross Margin': '_grossMargin',
          'Operating Income': 'operatingIncome',
          '{%} Operating Margin': '_operatingMargin',
          'Net Income': 'netIncome',
          '{%} Net Margin': '_netMargin',
      },
      properties: {
        'title': 'Historical Values',
        'currency': currency,
        'number_format': 'M',
        'display_averages': true,
        'column_order': 'descending',
      },
    });
    
    // Historical growth rates table
    historical_computed_data.renderTable({
      start_date: startDate,
      data: {
          'Revenue': 'revenue',
          '{%} Revenue Growth Rate': '_revenueGrowthRate',
          'Net Income': 'netIncome',
          '{%} Net Margin': '_netMargin',
          '{%} Net Income Growth Rate': '_netIncomeGrowthRate',
          'Stockholders Equity': 'totalStockholdersEquity',
          '{%} Equity Growth Rate': '_totalStockholdersEquityGrowthRate',
          '{%} Return on Invested Capital (ROIC)': '_roic',
          'After-tax Operating Income': 'nopat',
          '{%} Income Tax Rate': '_taxRate',
          'Invested Capital': 'investedCapital',
      },
      properties: {
        'title': 'Historical Growth Rates',
        'currency': currency,
        'number_format': 'M',
        'display_averages': true,
        'column_order': 'descending',
      },
    });
    // +---------------- END OF TABLES SECTION ---------------+
  }
  catch (error) {
    throwError(error);
  }
});

Description(`<h5>Discounted Future Market Cap</h5>
	This model estimates the intrinsic value of a common share by projecting the Future Market Capitalization using the estimated PE Ratio and then discounting it to the Present using an Annual Required Rate of Return.
`,
  {
    _DISCOUNT_RATE: [
      '{Equation} \\text{Discount Rate} = \\text{Cost of Equity} = \\text{Risk Free Rate} + \\text{Beta} * \\text{Market Premium}',
      '{Paragraph} The cost of equity is the theoretical rate of return that an equity investment should generate. It is calculated using the CAPM formula.',
      '{Link} https://www.investopedia.com/terms/c/costofequity.asp#mntl-sc-block_1-0-20',
    ],
    HISTORICAL_YEARS: 'The number of historical years used to calculate averages for historical data.',
    PROJECTION_YEARS: 'The number of years for projecting the analysis into the future.',
    _REVENUE_GROWTH_RATE: `The annual revenue growth rate is applied to projected revenue starting from the second projection year onward.`,
    _NET_INCOME_MARGIN: `Net Income expressed as a percentage of Revenue.`,
    PE_RATIO: [
        `Estimated Price to Earnings (PE) Ratio at the end of the projection period.`,
        '{Equation} \\text{Future Market Cap} = \\text{PE Ratio} * \\text{Projected Net Income}',
    ],
    _RISK_FREE_RATE: 'The risk-free rate represents the interest an investor would expect from an absolutely risk-free investment over a specified period of time.'+
    ' By default, it is equal to the current yield of the U.S. 10 Year Treasury Bond.',
    _MARKET_PREMIUM: 'Market risk premium represents the excess returns over the risk-free rate that investors expect for taking on the incremental risks connected to the equities market.',
    BETA: 'Beta is a value that measures the price fluctuations (volatility) of a stock with respect to fluctuations in the overall stock market.',
  });

File: source-code/valuations/discounted-future-market-cap.py
"""
    Model: Discounted Future Market Cap

    © Copyright discountingcashflows.com
"""

# Render model description
model.render_description(r"""
## Discounted Future Market Cap

This model estimates the intrinsic value of a common share by projecting the Future Market Capitalization using the estimated PE Ratio and then discounting it to the Present using an Annual Required Rate of Return.
""")

# Initialize assumptions
assumptions.init({
    "data": {
        "%discount_rate": None,
        "beta": data.get("profile:beta", default=1),
        "%risk_free_rate": data.get("treasury:year10"),
        "%market_premium": data.get("risk:totalEquityRiskPremium"),
        "pe_ratio": data.get("ratio:priceEarningsRatio"),
        "%revenue_growth_rate": None,
        "%net_income_margin": None,
        "historical_years": 10,
        "projection_years": 5,
    },
    "hierarchies": [{
        "parent": "%discount_rate",
        "children": ["beta", "%risk_free_rate", "%market_premium"]
    }]
})

# Calculate cost of equity
cost_of_equity = assumptions.get("%risk_free_rate") + assumptions.get("beta") * assumptions.get("%market_premium")
assumptions.set("%discount_rate", cost_of_equity)

# Compute metrics
data.compute({
    "%grossMargin": "income:grossProfit / income:revenue",
    "%operating_margin": "income:operatingIncome / income:revenue",
    "%netMargin": "income:netIncome / income:revenue",
    "%revenueGrowthRate": "function:growth:income:revenue",
    "%netIncomeGrowthRate": "function:growth:income:netIncome",
    "%totalStockholdersEquityGrowthRate": "function:growth:balance:totalStockholdersEquity",

    # Calculating NOPAT (Net Operating Profit After Taxes)
    "%taxRate": "income:incomeTaxExpense / income:incomeBeforeTax",
    "incomeTax": "income:operatingIncome * %taxRate",
    "nopat": "income:operatingIncome - incomeTax",

    # Calculating Invested Capital
    "investedCapital": [
        "balance:totalNonCurrentAssets", "-", "balance:cashAndCashEquivalents", "+",
        "balance:totalCurrentAssets", "-", "balance:totalCurrentLiabilities"
    ],
    "%roic": "nopat / investedCapital",
})

data.set_default_range(f"{-assumptions.get('historical_years')}->0")

assumptions.set("%net_income_margin", data.average(f"%netMargin"))
assumptions.set_bounds(
    "%net_income_margin", low=0, high="100%"
)

assumptions.set("%revenue_growth_rate", data.average(f"%revenueGrowthRate"))
assumptions.set_bounds(
    "%revenue_growth_rate", low=0, high="25%"
)

# Compute forecasted values and ratios
data.compute({
    "income:revenue": f"""
        income:revenue:-1 * 
        (1 + {assumptions.get('%revenue_growth_rate')})
    """,
    "%revenueGrowthRate": "function:growth:income:revenue",
    "income:netIncome": f"income:revenue * {assumptions.get('%net_income_margin')}",
}, forecast=assumptions.get("projection_years"))

projected_revenue = data.get("revenue")
projected_net_income = data.get(f"income:netIncome:{assumptions.get('projection_years')}")

shares_outstanding = data.get("income:weightedAverageShsOut")
future_market_cap = assumptions.get("pe_ratio") * projected_net_income
discounted_future_market_cap = future_market_cap / (
            (1 + assumptions.get("%discount_rate")) ** assumptions.get("projection_years"))

stock_value = discounted_future_market_cap / shares_outstanding

# Set the final value
model.set_final_value({
    "value": stock_value,
    "units": "$"
})

# Render results
model.render_results([
    [stock_value, "Present Value", "$"],
    [projected_net_income, "Estimated net income", "$"],
    [future_market_cap, "Estimated market capitalization", "$"],
    [discounted_future_market_cap, "Market capitalization discounted to present", "$"],
    [shares_outstanding, "Shares Outstanding", "#"],
    [data.get("income:eps"), "Earnings Per Share (EPS)", "$"],
    [data.get("profile:price"), "Market Price", "$"],
    [data.get("ratio:priceEarningsRatio"), "Price to Earnings (PE) Ratio", "#"]
])

# Render chart
model.render_chart({
    "data": {
        "income:revenue": "Revenue",
        "income:netIncome": "Net Income"
    },
    "start": -assumptions.get("historical_years"),
    "end": "*",
    "properties": {
        "title": "Historical and forecasted data",
        "set_editable": [
            "income:revenue",
            "income:netIncome"
        ]
    }
})

# Render table for estimated future values
model.render_table({
    "data": {
        "income:revenue": "Revenue",
        "%revenueGrowthRate": "Revenue Growth Rate",
        "income:netIncome": "Net Income",
    },
    "start": -1,
    "end": "*",
    "properties": {
        "title": "Estimated Future Values",
        "order": "ascending",
        "number_format": "M"
    }
})

# Render historical values table
model.render_table({
    "data": {
        "income:revenue": "Revenue",
        "income:costOfRevenue": "Cost of Revenue",
        "income:grossProfit": "Gross Profit",
        "%grossMargin": "Gross Margin",
        "income:operatingIncome": "Operating Income",
        "%operating_margin": "Operating Margin",
        "income:netIncome": "Net Income",
        "%netMargin": "Net Margin",
    },
    "start": -assumptions.get("historical_years"),
    "end": 0,
    "properties": {
        "title": "Historical Values",
        "display_averages": True,
        "order": "descending",
        "number_format": "M"
    },
})

# Render historical growth rates table
model.render_table({
    "data": {
        "income:revenue": "Revenue",
        "%revenueGrowthRate": "Revenue Growth Rate",
        "income:netIncome": "Net Income",
        "%netMargin": "Net Margin",
        "%netIncomeGrowthRate": "Net Income Growth Rate",
        "balance:totalStockholdersEquity": "Stockholders Equity",
        "%totalStockholdersEquityGrowthRate": "Equity Growth Rate",
        "%roic": "Return on Invested Capital (ROIC)",
        "nopat": "After-tax Operating Income",
        "%taxRate": "Income Tax Rate",
        "investedCapital": "Invested Capital",
    },
    "start": -assumptions.get("historical_years"),
    "end": 0,
    "properties": {
        "title": "Historical Growth Rates",
        "display_averages": True,
        "order": "descending",
        "number_format": "M"
    },
})

# Set description for assumptions
assumptions.set_description({
    "%discount_rate": r"""
        ## Discount Rate

        `Discount Rate` = `Cost of Equity` = `Risk Free Rate` + `Beta` * `Market Premium`

        The cost of equity is the theoretical rate of return that an equity investment should generate. It is calculated using the CAPM formula.

        [Read More](https://www.investopedia.com/terms/c/costofequity.asp#mntl-sc-block_1-0-20)
    """,

    "beta": r"""
        ## Beta

        Beta is a value that measures the price fluctuations (volatility) of a stock with respect to fluctuations in the overall stock market.
    """,

    "%risk_free_rate": r"""
        ## Risk-Free Rate

        The risk-free rate represents the interest an investor would expect from an absolutely risk-free investment over a specified period of time.
        By default, it is equal to the current yield of the U.S. 10 Year Treasury Bond.
    """,

    "%market_premium": r"""
        ## Market Premium

        Market risk premium represents the excess returns over the risk-free rate that investors expect for taking on the incremental risks connected to the equities market.
    """,

    "pe_ratio": r"""
        ## Price to Earnings (PE) Ratio

        Estimated Price to Earnings (PE) Ratio at the end of the projection period.

        `Future Market Cap` = `PE Ratio` * `Projected Net Income`
    """,

    "%revenue_growth_rate": r"""
        ## Revenue Growth Rate

        The annual revenue growth rate is applied to projected revenue starting from the second projection year onward.
    """,

    "%net_income_margin": r"""
        ## Net Income Margin

        Net Income expressed as a percentage of Revenue.
    """,

    "historical_years": r"""
        ## Historical Years

        The number of historical years used to calculate averages for historical data.
    """,

    "projection_years": r"""
        ## Projection Years

        The number of years for projecting the analysis into the future.
    """,
})

print("End of model.")
